### PR TITLE
[Refactor] Passwort Hash/Set

### PR DESCRIPTION
**Pull request by 6543**
* Rename HashPassword() to SetPassword
* Generate new Salt on password set
* Drop Passwords witch are Valid with "" string and use empty passwd field for IsPasswordSet() func

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -395,10 +395,16 @@ func hashPassword(passwd, salt, algo string) string {
...
+// change passwd, salt and passwd_hash_algo fields
+func (u *User) SetPassword(passwd string) (err error) {
+	if u.Salt, err = GetUserSalt(); err != nil {
+		return err
+	}
 	u.PasswdHashAlgo = setting.PasswordHashAlgo
 	u.Passwd = hashPassword(passwd, u.Salt, setting.PasswordHashAlgo)
+
+	return nil
 }
```
Comment (zeripath) [inline models/user.go]: This doesn't appear to prevent empty passwords being set.

**Review by zeripath — CHANGES_REQUESTED** [review]
This doesn't appear to

**Comment by zeripath** [comment]
You'll need to check every call to CreateUser and ensure that an empty password cannot be set - for example this code at present will set "" as the password for external users and makes it impossible to detect if a user is an external user through ispasswordset

**Comment by 6543** [comment]
@zeripath why not use the SetPasswort function to check this :point_up: :)

File: models/user.go [hunk 2]
```diff
@@ -395,10 +395,20 @@ func hashPassword(passwd, salt, algo string) string {
 	return fmt.Sprintf("%x", tempPasswd)
 }
 
-// HashPassword hashes a password using the algorithm defined in the config value of PASSWORD_HASH_ALGO.
-func (u *User) HashPassword(passwd string) {
+// SetPassword hashes a password using the algorithm defined in the config value of PASSWORD_HASH_ALGO
+// change passwd, salt and passwd_hash_algo fields
+func (u *User) SetPassword(passwd string) (err error) {
+	if len(passwd) == 0 {
+		return fmt.Errorf("no passord to be set")
```
Comment (zeripath) [inline models/user.go]: This is going to break creation of external users

Comment (6543) [reply models/user.go]: why is SetPassword used on external Users?!?

Comment (zeripath) [reply models/user.go]: It's in:

`func CreateUser(u *User) (err error) {`

Comment (6543) [reply models/user.go]: right this is independent ...

Comment (6543) [reply models/user.go]: @zeripath fixed - still I think before mrege/lgtm we neet to test this a bit ...

Comment (6543) [reply models/user.go]: (I'm worried about edge-cases)

Comment (zeripath) [reply models/user.go]: Could do something like test the LoginSource and if it's == 0 then assert the password has to be something non-empty - but setting password to empty could just be used to disable login locally I guess.

Comment (6543) [reply models/user.go]: Oauth2 account creation works :)
local account too :rocket: 

not tested: LDAP, SMTP, ...

Comment (zeripath) [reply models/user.go]: I think we should allow setPassword to set to "" - logic disallowing users to set their own password to "" belongs at the router level.

The reasoning here is that setPassword appears to be the only way to change the password - and therefore disable local login.

Comment (6543) [reply models/user.go]: hmm so if it is  "" we just clear all pwd fileds and it should be fine ...

**Review by zeripath — APPROVED** [review]

**Review by a1012112796 — APPROVED** [review]

File: routers/admin/users.go [hunk 3]
```diff
@@ -267,7 +267,10 @@ func EditUserPost(ctx *context.Context, form auth.AdminEditUserForm) {
 			ctx.ServerError("UpdateUser", err)
 			return
 		}
-		u.HashPassword(form.Password)
+		if err = u.SetPassword(form.Password); err != nil {
+			ctx.InternalServerError(err)
```
Comment (lunny) [inline routers/admin/users.go:271]: `ServerError` is the correct function but not `InternalServerError`

Comment (6543) [reply routers/admin/users.go:271]: @lunny this should be not issue - It's used else where too:
* https://github.com/go-gitea/gitea/pull/14282/files#diff-e9dab0c829326d553cbf4a8c95d2f55d7b2541317aec8018b9e247d79352bfd9L285
* https://github.com/go-gitea/gitea/pull/14282/files#diff-e9dab0c829326d553cbf4a8c95d2f55d7b2541317aec8018b9e247d79352bfd9R293

and is only a shortcout for ServerError ...

Comment (6543) [reply routers/admin/users.go:271]: @lunny but here it is: https://github.com/go-gitea/gitea/pull/14341

**Pull request merged**